### DISCUSSION TITLE
Dealing with Deadlocks/Node Downtimes?

### DISCUSSION THREAD

**Discussion by andrewm-aero** [rook/rook#9678] [Q&A]
This is more of a "How are you solving this?" rather than "How to do X with rook?".

One issue we've frequently encountered is that any sort of network disruption to a node can quickly bring down any apps with persistent storage on that node, or their dependencies, as any RBD volumes attached to that node are stuck in a purgatory where they aren't being used by any pods, but they can't be attached to new pods until that node's kubelet and csi plugins resume to say to kubernetes and ceph "no, really, its okay, these aren't being used anymore, go ahead and let other nodes/pods use them". We also occasionally see what we believe to be a kernel deadlock where RBD-consuming processes get stuck in uninterruptable sleep, requiring a hard power cycle of the host in question, risking data loss.

We mainly went with rook/ceph as a way to protect against single-point failures with storage on individual drives and hosts, but these issues has shown us that this has just kicked that can down the road, and now any node consuming RBDs is a single point of failure. There's a few ways we've tried mitigating this, like using replicated storage (e.g. leader/follower pairs of postgres databases), but this means that we're now using 6x (or more) the storage instead of just 3x, which isn't great, and doesn't account for apps which stubbornly refuse to implement any sort of leader-follower or quorum based HA (e.g. Jenkins).

Has anyone else encountered similar issues? If so, how have you prevented/mitigated it? Are we just doing something silly?

**Comment by travisn** [comment]
This is certainly a difficult problem and it's not unique to anything you're doing. This is an issue long tracked in #1507 and a workaround documented [here](https://rook.github.io/docs/rook/latest/ceph-csi-troubleshooting.html#node-loss). 

Since this is an issue more general than just rook, there is also a proposal to the CSI spec in progress: https://github.com/container-storage-interface/spec/pull/477

The [Poison Pill operator](https://operatorhub.io/operator/poison-pill-operator) is a solution that I've seen more recently, though not sure if it's ready for production.

**Reply by Madhu-1** [reply]
@travisn poison pill works only in a few cases and it has some pre-req for machine health checks,  won't work with all kubernetes clusters? going forward is the CSI proposal you mentioned here it might take some time, the manual workaround we already document is the only possible way for now.

**Comment by travisn** [comment]
CephFS volumes also don't suffer from this issue, but then of course you lose some performance compared to rbd.

@Madhu-1 @humblec Other thoughts on dealing with the node loss for rbd volumes?

**Reply by andrewm-aero** [reply]
That's the conclusion my team reached as well, so its good to know we're not off in left field. However, that raises the perhaps uncomfortable question: What is the point of RBD?

In my team's experience, persistent apps fall into one a few categories:
* Multiple apps which require concurrent access to the same files (We find a lot of old legacy business apps work like this) - CephFS makes perfect sense.
* Apps which require object storage (a lot of newer apps) - RGW is a match made in heaven.
* Apps which don't have any HA solution (Jenkins) - CephFS is required for any semblance of high availability, and using RBD is a liability.
* Apps which perform their own replication via leader/follower (e.g. postgres) or a quorum (zookeeper, kafka, elasticsearch) - OpenEBS and similar make more sense, since it lets you allocate access to "bare metal" drives. You "could" use RBD (and we still do), but its a lot of overhead for (what we perceive as) little value. What would be brilliant is if we could provision multiple PVCs for those replicas, and somehow communicate to ceph that their PG's cannot be put on the same CRUSH host, which would allow us to only have one ceph PG replica per PVC replica, removing the overhead, but still guaranteeing that the app can survive a storage node loss.

There's basically only two use cases I can see where RBD makes sense:
* Where HA doesn't matter (rapid prototyping, perhaps)
* Your pods are already pinned to a specific node (perhaps a GPU, FPGA, or other exotic peripheral is required for your workload), so if that node goes down, your app isn't going to be running anyways. If that's the case, then your node is "special" anyways, and it just makes sense to give it bare metal storage access on RAID or something.

Any thoughts?

**Reply by leseb** [reply]
RBD has a few more advantages like:

* Performance over CephFS
* Copy on write/read clones
* Native built-in encryption (not used by Ceph-CSI) but still encryption with cryptsetup is used so still a plus over CephFS
* Ressource requirements on the cluster, you don't need any CephFS so a lot of memory is saved
* More robust implementation of the block mirroring where the CephFS mirror is quite new
* Raw disks can be used for Hypervisors if needed (would still prefer libvirt plugin though)
* RBD quota are better supported than Subvolumegroup ones (still in-dev)
* More things I'm probably forgetting :)

**Reply by andrewm-aero** [reply]
Sure, I get that RBD has more features and performance than CephFS, that's a given. My train of thought is that if you need HA, then RBD isn't an option, since it has a single point of failure (the node mounting the RBD volume), and if HA isn't a requirement, or provided through some other means (application-specific data redundancy) then what benefits does using RBD in the first place provide, compared to using local-type persistent volumes, statically provisioned or otherwise?

Maybe its just my own personal inexperience, but if you're using networked storage, to me that means either you want the data to be resilient (i.e. no single points of failure), or its not reasonable to put all of the bits on the same box, in which case you're going to want it to be resilient anyways, because that's an awful lot of data to go unavailable or missing due to a single failure.

I guess I'm struggling to think of a use where you want/need those (very cool and useful) features, but where a single node failure halting your operations is acceptable.

I hope I'm not sounding like I'm trying to badmouth Ceph or Rook, both are fantastic tools, and my life would be much more difficult without them, but the experiences leading up to this thread have left me questioning if I should ever be recommending RBD to my customers and users.
